### PR TITLE
fix: use correct file uri handling

### LINKED ISSUE #277: Version 1.7.1 can’t handle some paths in Nova 9.x
PR #276 addressed changes of the API contract for file paths with Nova’s update to version 9. Unluckily, due to  faulty analysis of the cause underlying these changes, the resulting patch is bound to cause problems down the road. What Nova 9 did change is not URI handling as such. What changed is that [absolute paths returned by the Nova API do not start at the notional volume mount point (“/Volumes/Macintosh HD” by default) anymore, but at the \*nix root](https://devforum.nova.app/t/suggestions-for-converting-workspace-path/465/16). This also translates to document URIs.

The problem with #276 is that, probably due to this faulty analysis, it throws the baby out with the bathwater by removing the URI decoding step. Document URIs are still URIs in Nova 9, i.e. they still do encode non URI fragment characters like spaces and, without the decode step, version 1.7.1 will fail on all paths containing such characters.

PS: good to see you are still around, @apexskier!

### LINKED ISSUE #279: ESLint fails on all my projects (working on VSCode)
ESLint doesn't work at all. All the projects I tested is working out of the box on VSCode with ESLint extension.

**Environment**

- Nova version: `9.1 build 383286`
- Extension version: `1.7.2`
- ESLint version: `8.11.0`
- macOS version: `12.2.1`
- NodeJS information:
  - node version: `17.7.1`
  - npm version: `8.5.2`
  - installation method: homebrew

**Extension console output**

```sh
ESLint[9:38:48.428000] Failed to lint

ESLint[9:38:48.428000]   stderr:  
Oops! Something went wrong! :(

ESLint: 8.11.0

No files matching the pattern "/Code/{client-name}/{project-name}/webapp/src/App.tsx" were found.
Please check for typing mistakes in the pattern.



ESLint[9:38:48.428000]   command:  /Users/gabriele/Code/{client-name}/{project-name}/webapp/node_modules/.bin/eslint

ESLint[9:38:48.428000]   args:  /Code/{client-name}/{project-name}/webapp/src/App.tsx --fix --format=json

ESLint[9:38:48.429000] Extension encountered an uncaught exception:
/Users/gabriele/Library/Application Support/Nova/Extensions/apexskier.eslint/Scripts/main.dist.js (Line 372, Column 32)
Error: failed to lint (2)
    @file:///Users/gabriele/Library/Application%20Support/Nova/Extensions/apexskier.eslint/Scripts/main.dist.js:372:32
    @[native code]
    promiseReactionJob@[native code]

```

**Additional context**

I opened the folder `/Users/gabriele/Code/{client-name}/{project-name}/webapp` in Nova.

I'm a bit worried about this line:

```
No files matching the pattern "/Code/{client-name}/{project-name}/webapp/src/App.tsx" were found.
```

Because that seems an absolute path and it shouldn't. The correct path is `/Users/gabriele/Code/{client-name}/{project-name}/webappsrc/App.tsx`

### LINKED ISSUE #282: Import is reserved keyword
Hey!

Trying to use plugin - gives me that `import is reserved` in the Nova, but running eslint separately on any file works fine - without errors.

**To reproduce**

1. Open any file with `import` declaration

Or

1. Create new TS file
2. Add import declaration - no errors
3. Save it - after save - error appears

**Expected behavior**

Should follow the output of the eslint run from cli

**Screenshots**

image

**Environment**

- Nova version: 9.1 383286
- Extension version: 1.7.2
- ESLint version: v7.18.0
- macOS version: 12.3 (21E230)
- NodeJS information:
  - node version: v14.19.0
  - npm version: 6.14.16
  - installation method: dmg file

**Extension console output**

```
ESLint[20:10:27.512000] activating...
ESLint[20:10:27.691000] activated
```

### LINKED ISSUE #293: ESLint fails on a project reporting it can't find a configuration file
ESLint Nova was working on that project a few weeks ago, now it doesn't work at all. Running ESLint in Terminal works fine.

**Environment**

- Nova version: `9.1 Build 383286`
- Extension version: `1.7.2`
- ESLint version: `8.12.0`
- macOS version: `12.3.1 (21E258)`
- NodeJS information:
  - node version: `v16.14.0`
  - npm version: `8.3.1`
  - installation method: `nvm`

**Extension console output**

```
ESLint[14:30:01.552000] 
Oops! Something went wrong! :(

ESLint: 8.12.0

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config

ESLint looked for configuration files in /Developer/gated-art-web/components and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://eslint.org/chat/help



ESLint[14:30:01.553000] Extension encountered an uncaught exception:
/Users/alex/Library/Application Support/Nova/Extensions/apexskier.eslint/Scripts/main.dist.js (Line 306, Column 28)
Error: failed to get eslint config for /Developer/gated-art-web/components/TimelinePostContextMenu.vue: 2
    @file:///Users/alex/Library/Application%20Support/Nova/Extensions/apexskier.eslint/Scripts/main.dist.js:306:28
    @[native code]
    promiseReactionJob@[native code]
```

**Additional context**

I use `.eslintrc.yml` as my config file for the project and it's located in the project root. ESLint runs without any problem when invoked from command line.

### LINKED ISSUE #294: fix on save  gives error
1. Set an Config file and set auto fix auf enable
2. Open a *.js File 
3. Save file, error comes up

Settings
Bildschirmfoto 2022-04-07 um 21 52 39

Error
Bildschirmfoto 2022-04-07 um 21 51 45

**Environment**

- Nova version: 9.1 rev 383286
- Extension version: 1.7.2
- ESLint version: 8.12.0
- macOS version: 12.3.1
- NodeJS information:
  - node version: v17.7.2 
  - npm version: 8.5.2
  - installation method: asdf nodejs plugin

Hi!
the problem with the plugin is i cant use the fix on save option. Everytime i run in an error (screenshot above). Is the problem by me or the plugin ?

Greetings ! 
Jim

### PR DESCRIPTION
**Pull request by gnattu**
This pr includes the correct file path handling for nova version <9 and >=9. Fixes #277, fixes #282, probably related to #280 as well. Most of the uri normalization implementation is done by @kopischke published on nova dev forum.

### CODE REVIEW

**Comment by kopischke** [comment]
Most certainly related to #280 IMO (I predicted something like that would happen in #277: spaces are among the encoded characters in URIs, i.e. `%20`).

**Comment by kopischke** [comment]
FWIW, I just added a `decodeFileURI()` function (hopefully) resistant to `decodeURI()` API contract changes to the library the above functions originally come from. No link yet, sorry (the _NovaNova_ tooling and library extension these are part of is still very much WIP), but here is the current code:

```js
/**
 * Convert a “file://” URI into a a path in a way that conforms to RFC 8089
 * and is resistant to contract changes of Nova’s “decodeURI()” implementation.
 * @see https://en.wikipedia.org/wiki/File_URI_scheme
 * @param {string} uri - The URI to convert.
 * @returns {?string} The path, if uri was a file URI.
 */
function decodeFileURI (uri) {
  const regex = new RegExp('^file:(//[^/]+)?(?=/)')
  const fileURI = uri.replace(regex, '')
  if (fileURI !== uri) {
    const parts = nova.path.split(fileURI).map(part => decodeURI(part))
    return nova.path.join(...parts)
  }
}
```

**EDIT:** updated to conform to RFC 8089.

Combine this with `nixalise()` like so
```js
const cleanPath = path ? nixalize(decodeFileURI(path)) : null;
```
and Björn Stronginthearm is your uncle.

EDIT: I have just noticed I have omitted all context. Find it here: #282 ([comment](https://github.com/apexskier/nova-eslint/issues/282#issuecomment-1078889213)).

**Comment by apexskier** [comment]
Thanks! Looks like there will be a few linting errors after the build runs - could you also [add a line to the changelog](https://github.com/apexskier/nova-eslint/blob/main/CONTRIBUTING.md#changelog) (I keep forgetting to request this).

I'm going to pull this down and do some more extensive testing, but it'll take me a bit to get to it.

**Comment by drwpow** [comment]
It’s a shame we can’t use Node’s built-in [URL handling](https://nodejs.org/api/url.html#urlfileurltopathurl) which takes care of all of this. I’m still new to Nova / Nova Extensions but is there no way to access Node builtins from extensions?

File: src/process.ts [hunk 1]
```diff
@@ -221,7 +257,7 @@ export function runLintPass(
   const eslint = eslintPath;
   // remove file:/Volumes/Macintosh HD from uri
   const cleanPath = path
-    ? "/" + decodeURI(path).split("/").slice(5).join("/")
+    ? "/" + decodeURI(nixalize(path)).split("/").slice(1).join("/")
```
Comment (drwpow) [inline src/process.ts]: This fix works on my machine! 💯 Thank you 🙏

**Comment by kopischke** [comment]
> It’s a shame we can’t use Node’s built-in [URL handling](https://nodejs.org/api/url.html#urlfileurltopathurl) which takes care of all of this. I’m still new to Nova / Nova Extensions but is there no way to access Node builtins from extensions?

As Cameron stated [elsewhere](https://github.com/apexskier/nova-eslint/issues/282#issuecomment-1083810726), Nova extensions run in Nova’s own, [JavaScriptCore](https://developer.apple.com/documentation/javascriptcore) based runtime. [The only way to use Node modules is to bundle them (via Webpack, Rollup or such), and anything that relies on Node builtins is out](https://library.panic.com/nova/npm-packages-in-extensions/).

In the specific case of file URL handling, that is not too dramatic: the platform dependent functionality of the Node URL module is not needed (Nova is a macOS only editor; AFAICS, correctly decoding Windows UNC paths should not be a use case), and the POSIX only functionality can be encapsulated in a simple function [like my `decodeFileURI()` above](https://github.com/apexskier/nova-eslint/pull/283#issuecomment-1078962694).

**Comment by gnattu** [comment]
I updated the branch with lint fixes, using new `decodeFileURI ` and updated changelog.

**Comment by davidbonnet** [comment]
For the curious, the discussion on Panic's forum mentioned above: https://devforum.nova.app/t/suggestions-for-converting-workspace-path/465

**Comment by jclow-sift** [comment]
Is there anything that interested non-maintainers can do to help get this moving? ESLint integration being broken is a real drag, so I am motivated (and I'm sure others are too) to help get this updated

**Comment by apexskier** [comment]
Yes, there is! First, sorry for my inactivity here, I'm hoping to have more time to contribute here next month.

If anyone wants to help push this along, the main thing (aside from directly contributing PRs) you can do to help is to run this branch locally and check the fix, both on Nova 9 and older versions. [There are some basic instructions about doing this in the contribution docs.](https://github.com/apexskier/nova-eslint/blob/main/CONTRIBUTING.md#running-locally)

**Comment by eaviles** [comment]
@apexskier I just confirmed the changes in this PR as working in Nova 9.3 build 396482 with ESLint 8.14.0. Let me know if you need anything from my side. Happy to help verifying this.

**Comment by kpitt** [comment]
I was experiencing the issue described in #293. I just tested this PR locally and I am now getting ESLint results, so I can confirm that this fixes the issue for me. I'm running Nova 9.3, Build 396482 on macOS 12.1 with ESLint 7.24.0.

**Comment by mindgrub-rkoch** [comment]
Is there an ETA on when this will be merged in?

**Comment by apexskier** [comment]
Apologies for my leave of absence, I finally have some time to work on my extensions.

**Comment by apexskier** [comment]
Included in release [v1.7.3](https://github.com/apexskier/nova-eslint/releases/tag/v1.7.3)

**Pull request merged**